---
title: "546x"
output: html_document
---


# *Author: Jose Lopez*


# Part I

### Inpecting the data
```{r}
genotypes <- read.csv("fang_et_al_genotypes.txt", sep="\t")
snp <- read.csv("snp_position.txt", sep="\t")

str(genotypes)
nrow(genotypes)
ncol(genotypes)
dim(genotypes)



str(snp)
nrow(snp)
ncol(snp)
dim(snp)



file.size("fang_et_al_genotypes.txt")
file.size("snp_position.txt")
```


### Separate the data Maize: "ZMMIL,ZMMLR,ZMMMR" and Teosinte: "ZMMPBA ZMPIL ZMPJA"
```{r}
Maize <- genotypes[genotypes$Group=="ZMMIL" | genotypes$Group=="ZMMLR" | genotypes$Group=="ZMMMR", ]
teosinte <- genotypes[genotypes$Group=="ZMPBA" | genotypes$Group=="ZMPIL" | genotypes$Group=="ZMPJA", ]
```


### transpose the files in one column

```{r}
transposed_teosinte <- as.data.frame(t(teosinte[,-1]))
colnames(transposed_teosinte) <- teosinte$Sample_ID

transposed_Maize <- as.data.frame(t(Maize[,-1]))
colnames(transposed_Maize) <- Maize$Sample_ID
```

### Create the data frame with colums required for the assignment 

```{r}
snp_id_chr_pos <-snp[,c(1,3,4)]
```

### join the transposed files with the snp_id_chr_pos dataframe and remove the rows that have "unkown and multiple positions"

```{r}
Joint_teosinte <- merge(x = snp_id_chr_pos, y = transposed_teosinte, by.x = "SNP_ID", by.y ="row.names", all.y = TRUE)

Joint_Maize <- merge(x = snp_id_chr_pos, y = transposed_Maize, by.x = "SNP_ID", by.y ="row.names", all.y = TRUE)


Joint_teosinte$Position <- as.numeric(Joint_teosinte$Position)
Joint_teosinte$Chromosome <- as.numeric(Joint_teosinte$Chromosome)


Joint_teosinte <- Joint_teosinte[!(Joint_teosinte$Position== "unknown") | !(Joint_teosinte$Chromosome== "unknown") | !(Joint_teosinte$Chromosome== "multiple"),]
Joint_teosinte <- Joint_teosinte[!(is.na(Joint_teosinte$Position)),]


Joint_Maize$Position <- as.numeric(Joint_Maize$Position)
Joint_Maize$Chromosome <- as.numeric(Joint_Maize$Chromosome)


Joint_Maize <- Joint_Maize[!(Joint_Maize$Position== "unknown") | !(Joint_Maize$Chromosome== "unknown") | !(Joint_Maize$Chromosome== "multiple"),]
Joint_Maize <- Joint_Maize[!(is.na(Joint_Maize$Position)),]

```


### Now we can sort the files by position:

```{r}

Joint_teosinte <- Joint_teosinte[order(Joint_teosinte$Position),]

Joint_Maize <- Joint_Maize[order(Joint_Maize$Position),]

```

### After this, we can split the dataframes according to the chromosome number and save the corresponding files

```{r}

teosinte_split <- split(Joint_teosinte, Joint_teosinte$Chromosome)

teosinte_Chromosome_1 <- teosinte_split$`1`
teosinte_Chromosome_2 <- teosinte_split$`2` 
teosinte_Chromosome_3 <- teosinte_split$`3`
teosinte_Chromosome_4 <- teosinte_split$`4` 
teosinte_Chromosome_5 <- teosinte_split$`5`
teosinte_Chromosome_6 <- teosinte_split$`6` 
teosinte_Chromosome_7 <- teosinte_split$`7`
teosinte_Chromosome_8 <- teosinte_split$`8` 
teosinte_Chromosome_9 <- teosinte_split$`9`
teosinte_Chromosome_10 <- teosinte_split$`10` 

Maize_split <- split(Joint_Maize, Joint_Maize$Chromosome)


Maize_Chromosome_1 <- Maize_split$`1`
Maize_Chromosome_2 <- Maize_split$`2`
Maize_Chromosome_3 <- Maize_split$`3`
Maize_Chromosome_4 <- Maize_split$`4`
Maize_Chromosome_5 <- Maize_split$`5`
Maize_Chromosome_6 <- Maize_split$`6`
Maize_Chromosome_7 <- Maize_split$`7`
Maize_Chromosome_8 <- Maize_split$`8`
Maize_Chromosome_9 <- Maize_split$`9`
Maize_Chromosome_10 <- Maize_split$`10`



```

### now we replace the values ? with - and reverse the order

```{r}
Joint_teosinte_rev <- Joint_teosinte[order(Joint_teosinte$Position, decreasing=T),]

Joint_Maize_rev <- Joint_Maize[order(Joint_Maize$Position, decreasing=T),]


Joint_teosinte_rev[] <- lapply(Joint_teosinte_rev, as.character)
Joint_teosinte_rev[Joint_teosinte_rev == '?/?'] <- '-/-'

Joint_Maize_rev[] <- lapply(Joint_Maize_rev, as.character)
Joint_Maize_rev[Joint_Maize_rev == '?/?'] <- '-/-'

teosinte_split_rev <- split(Joint_teosinte_rev, Joint_teosinte_rev$Chromosome)

rev_teosinte_Chromosome_1 <- teosinte_split_rev$`1`
rev_teosinte_Chromosome_2 <- teosinte_split_rev$`2` 
rev_teosinte_Chromosome_3 <- teosinte_split_rev$`3`
rev_teosinte_Chromosome_4 <- teosinte_split_rev$`4` 
rev_teosinte_Chromosome_5 <- teosinte_split_rev$`5`
rev_teosinte_Chromosome_6 <- teosinte_split_rev$`6` 
rev_teosinte_Chromosome_7 <- teosinte_split_rev$`7`
rev_teosinte_Chromosome_8 <- teosinte_split_rev$`8` 
rev_teosinte_Chromosome_9 <- teosinte_split_rev$`9`
rev_teosinte_Chromosome_10 <- teosinte_split_rev$`10` 

Maize_split_rev <- split(Joint_Maize_rev, Joint_Maize_rev$Chromosome)

rev_Maize_Chromosome_1 <- Maize_split_rev$`1`
rev_Maize_Chromosome_2 <- Maize_split_rev$`2`
rev_Maize_Chromosome_3 <- Maize_split_rev$`3`
rev_Maize_Chromosome_4 <- Maize_split_rev$`4`
rev_Maize_Chromosome_5 <- Maize_split_rev$`5`
rev_Maize_Chromosome_6 <- Maize_split_rev$`6`
rev_Maize_Chromosome_7 <- Maize_split_rev$`7`
rev_Maize_Chromosome_8 <- Maize_split_rev$`8`
rev_Maize_Chromosome_9 <- Maize_split_rev$`9`
rev_Maize_Chromosome_10 <- Maize_split_rev$`10`
```



# Part II

### recreate the joint files

```{r}
transposed_genotypes <- as.data.frame(t(genotypes[,-1]))
colnames(transposed_genotypes) <- genotypes$Sample_ID

genotypes_joint <- merge(x = snp_id_chr_pos, y = transposed_genotypes, by.x = "SNP_ID", by.y ="row.names", all.y = TRUE)
```

### install the `reshape2` and `ggplot` package

```{r}
if (!require("reshape2")) install.packages("reshape2")
library("reshape2")
if (!require("ggplot2")) install.packages("ggplot2")
library(ggplot2)
if (!require("plyr")) install.packages("plyr")
library(plyr)
```

### plot the total number of SNPs in our dataset on each chromosome. 

```{r}
genotypes_joint$Chromosome <- factor(genotypes_joint$Chromosome, levels = c("1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "multiple", "unknown", "NA"))

Number_chromosome_plot <- ggplot(genotypes_joint, aes((Chromosome))) + geom_bar(stat=) + ggtitle("Number of SNP on each chromosome") +
        theme(
            axis.text = element_text(colour = "black"),
            axis.title.x = element_text(colour = "black", size=rel(3)),
            axis.title.y = element_text(colour = "blue", angle=45),
            panel.background = element_rect(fill="white"),
            panel.grid.minor.y = element_line(size=3),
            panel.grid.major = element_line(colour = "black"),
            plot.background = element_rect(fill="gray")
    )   

Number_chromosome_plot

```


### plot the in groups they belong 

```{r}
group_plot <- ggplot(genotypes, aes(Group)) +
  geom_bar() + ggtitle("                                                    Group") +
        theme(
            axis.text = element_text(colour = "black"),
            axis.title.x = element_text(colour = "black", size=rel(3)),
            axis.title.y = element_text(colour = "blue", angle=45),
            panel.background = element_rect(fill="white"),
            panel.grid.minor.y = element_line(size=3),
            panel.grid.major = element_line(colour = "black"),
            plot.background = element_rect(fill="gray")
    )   

group_plot

```

#### This will tell us What groups contribute most of these SNPs. The ZMMIL, ZMMLR and ZMPBA contribute the most to the SNP count.


### Melt the datasets using a vector with the names of all the SNPs, this will help us analyze the amount of heterozygicity we have.

```{r}
headers_names<- colnames(genotypes)[-c(1:3)]
melted_genotypes <- melt(genotypes, measure.vars = headers_names)
```

### assign all missing values as NA, Create a new columns to indicate whether a particular site is homozygous, has the same nucleotide on both chromosomes (i.e., A/A, C/C, G/G, T/T) or heterozygous. 

```{r}
melted_genotypes[ melted_genotypes == "?/?" ] = NA
melted_genotypes$isHomozygous <- (melted_genotypes$value=="A/A" | melted_genotypes$value=="C/C" | melted_genotypes$value=="G/G" | melted_genotypes$value=="T/T")
```


### Sort/Order the dataframe using Group and Species_ID values. 

```{r}
genotypes_sorted_by_ID <- melted_genotypes[order(melted_genotypes$Sample_ID),]
genotypes_sorted_by_Group <- melted_genotypes[order(melted_genotypes$Group),]
```

### Build a dataframe for homozygous, heterozygous NA, with Sample_ID. Melt the results and make a graph that shows homozygous, heterozygous, and missing proportions. 

```{r}
counting_ID <- ddply(genotypes_sorted_by_ID, c("Sample_ID"), summarise,counting_heterozygous=sum(!isHomozygous, na.rm=TRUE), counting_homozygous=sum(isHomozygous, na.rm=TRUE),  isNA=sum(is.na(isHomozygous)))

counting_ID_melt <- melt(counting_ID, measure.vars = c("counting_homozygous", "counting_heterozygous", "isNA"))

Missing_data_plot1 <- ggplot(counting_ID_melt,aes(x = Sample_ID, y= value, fill=variable)) + geom_bar(stat = "identity", position = "stack") + ggtitle("Missing data and amount of heterozygosity") +
        theme(
            axis.text = element_text(colour = "black"),
            axis.title.x = element_text(colour = "black", size=rel(3)),
            axis.title.y = element_text(colour = "blue", angle=45),
            panel.background = element_rect(fill="white"),
            panel.grid.minor.y = element_line(size=3),
            panel.grid.major = element_line(colour = "white"),
            plot.background = element_rect(fill="gray")
    )  + scale_fill_brewer(palette="Set1") 

Missing_data_plot1
```

#### The plot above shows how homozygous is bigger through all SNPs, and low for missing data. NOTE: There are a group of SNPs that do not have heterozygous alleles.




### Plot with  sorted groups
```{r}
counting_Group <- ddply(genotypes_sorted_by_Group, c("Group"), summarise, counting_homozygous=sum(isHomozygous, na.rm=TRUE), counting_heterozygous=sum(!isHomozygous, na.rm=TRUE), isNA=sum(is.na(isHomozygous)))

counting_Group_melt <- melt(counting_Group, measure.vars = c("counting_homozygous", "counting_heterozygous", "isNA"))

Missing_data_plot2 <-  ggplot(counting_Group_melt,aes(x = Group, y= value, fill=variable)) + geom_bar(stat = "identity", position = "stack") + ggtitle("Missing data and amount of heterozygosity ") +
        theme(
            axis.text = element_text(colour = "black"),
            axis.title.x = element_text(colour = "black", size=rel(3)),
            axis.title.y = element_text(colour = "blue", angle=45),
            panel.background = element_rect(fill="white"),
            panel.grid.minor.y = element_line(size=3),
            panel.grid.major = element_line(colour = "black"),
            plot.background = element_rect(fill="gray")
    )   + scale_fill_brewer(palette="Set1") 

Missing_data_plot2
```

#### The plot above shows groups that contribute the most to the SNP number and the number of heterozygous and homozygous.






### We will make a  scatter plot. 

```{r}
headers_names_joint<- colnames(genotypes_joint)[-c(1:3)]
melted_genotypes_joint <- melt(genotypes_joint, measure.vars = headers_names_joint)

melted_genotypes_joint[ melted_genotypes_joint == "?/?" ] = NA

melted_genotypes_joint$isHomozygous <- (melted_genotypes_joint$value=="A/A" | melted_genotypes_joint$value=="C/C" | melted_genotypes_joint$value=="G/G" | melted_genotypes_joint$value=="T/T")

genotypes_sorted_by_SNP <- melted_genotypes_joint[order(melted_genotypes_joint$SNP_ID),]


Observed_Het_per_locus <- ddply(genotypes_sorted_by_SNP, c("SNP_ID"), summarise, heterozygocity_count=sum(!isHomozygous, na.rm=TRUE), total_count=sum(!is.na(isHomozygous)))

Observed_Het_per_locus$Obs_heterozygocity <- (Observed_Het_per_locus$heterozygocity_count/Observed_Het_per_locus$total_count)

Observed_Het_per_locus_melt <- melt(Observed_Het_per_locus, measure.vars = "Obs_heterozygocity")

splot <- ggplot(Observed_Het_per_locus_melt,aes(x = SNP_ID, y= value, fill=variable)) + geom_point() +
        theme(
            axis.text = element_text(colour = "black"),
            axis.title.x = element_text(colour = "black", size=rel(3)),
            axis.title.y = element_text(colour = "blue", angle=45),
            panel.background = element_rect(fill="black"),
            panel.grid.minor.y = element_line(size=3),
            panel.grid.major = element_line(colour = "white"),
            plot.background = element_rect(fill="gray")
    ) +  scale_fill_brewer(palette="Set2") 
splot
```

#### This plot shows the ranges distribution of the observed heterozygocity for each SNP, being able to identify that the majority of the SNP had an observed heterosygocity between 0 and 0.5. One of the data points has a velue of 1.0 


